designs/odcr.md

```diff
@@ -0,0 +1,140 @@
+# On demand capacity reservation
+
+This documents proposes supporting ODCR in Karpenter
+
+## Background
+
+In AWS [ODCR](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-capacity-reservations.html) allows users to reserve compute capacity to mitigate against the risk of being 
+unabled to get on demand capacity. This is very helpful during seasonal holidays where higher traffic is expected. 
+
+### Capacity Reservations
+
+Each Capacity Reservation is defined with:
+
+- The Availability Zone in which to reserve the capacity
+- The number of instances for which to reserve capacity
+- The instance attributes, including the instance type, tenancy, and platform/OS
+- Instance match criteria
+  - Targeted -- only accept instances that matches all attributes + explicitly targeted the capacity reservation
+  - Open -- if capacity reservation accepts all instances that matches all attributes
+
+AWS also supports grouping Capacity Reservation into Capacity Reservation groups. 
+Both these entities are supported in Launch Template's CapacityReservationTarget [definitions](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-launchtemplate-capacityreservationtarget.html).
+
+## Goals
```

Comment:
One additional goal I might (strongly) suggest - the ability to define "fairness" in some form when an ODCR is shared amongst multiple nodepools/clusters. The scenario:
- An ODCR for 500 instances is created
- That ODCR is shared across x-number of AWS accounts (say 5)
- Each account contains a cluster that should be entitled to 1/x (roughly) percent of the reservation at minimum (each cluster at minimum is guaranteed access to launch 100 instances)

The most basic form could be simply providing a max/ceiling on the number of instances a nodepool can launch from a reservation and let users set that value. 

This is used in conjunction with the resource group - as more capacity is needed, more ODCRs are created and added into the "pool" of capacity, but that "pool" of capacity is shared amongst multiple Karpener nodepools across clusters/accounts

Reply:
Can't you already do this by having an ODCR NodePool with max vCPUs configured?

Reply:
> One additional goal I might (strongly) suggest - the ability to define "fairness" in some form when an ODCR is shared amongst multiple nodepools/clusters. The scenario:
> 
> * An ODCR for 500 instances is created
> * That ODCR is shared across x-number of AWS accounts (say 5)
> * Each account contains a cluster that should be entitled to 1/x (roughly) percent of the reservation at minimum (each cluster at minimum is guaranteed access to launch 100 instances)
> 
> The most basic form could be simply providing a max/ceiling on the number of instances a nodepool can launch from a reservation and let users set that value.
> 
> This is used in conjunction with the resource group - as more capacity is needed, more ODCRs are created and added into the "pool" of capacity, but that "pool" of capacity is shared amongst multiple Karpener nodepools across clusters/accounts

What you are describing is outside of the current capabilities of AWS itself, once those capabilities are available we can update or create another design document to incorporate this

Reply:
> Can't you already do this by having an ODCR NodePool with max vCPUs configured?

Agreed, seems like a quick win, but is management overhead for now

Reply:
> What you are describing is outside of the current capabilities of AWS itself

Which capability(s) exactly? You can share ODCRs across accounts

> Can't you already do this by having an ODCR NodePool with max vCPUs configured?

In theory, yes - but for most of the workloads that are utilizing GPUs via ODCRs, you aren't using vCPU as a proxy for your GPU allotment. p5s have double the vCPU capacity as p4s - so not a great proxy value

Not saying every use case has to be considered and implemented on the first go - but it seems we are missing the mark on ODCRs and how they are intended to be used:
- ODCRs should be added to a resource group and treated like a pool of compute resources. This pool of compute will expand/contract by adding/removing ODCRs
- ODCRs are not always 1:1 to a cluster/account/team (they rarely are 1:1) - again, its a pool of resources and that pool may need to be divided/shared amongst clusters/accounts/teams
- If you have a fixed pool of available resources, you will inevitably need to consider priority and fairness in terms of resource allocation. On day 1 you may have two teams who each consume half the pool by verbally agreeing to use x number of instances. As workloads grow in size, as is the norm for ML workloads, teams will need more resources and now you need to give the whole pool of resources to one team at a time but still ensure teams have equal time with the pool of resources

This is an RFC so not looking for every configuration or use case to be provided initially, but just want to ensure we are looking at the use cases and workloads and thinking about the implementation holistically - at least ensuring we have considered those common uses cases and can support them in the future without too much churn/disruption to the API

Reply:
Those cases were considered however we didn't choose the implement them on the first pass for a few reasons.

- Resource groups for capacity reservation makes the problem considerably more complicated because of its fallback logic to OD if the Capacity reservation group (CRG)'s ODCRs can't satisfy the request. This is complicated because it requires more complicated reconciliation because at this time we assume any ODCR will cost zero and have a fixed limit of instances. That isn't to say this won't be supported but to move this project forward we needed to narrow in a basic use case before we tackle the harder ones.
- Multi-account usage is supported in the selectors however you are right that we lack any fairness. Once again because we didn't get any early feedback from users we didn't implement any fairness between "shared" ODCRs.

Reply:
Please join Karpenter working group on June 6th 2PM PST to discuss (https://karpenter.sh/docs/contributing/working-group/)